src/operations/blas3/gemm_partial_local.hpp

```diff
@@ -285,160 +339,122 @@ class GemmPartial<input_t, output_t, DoubleBuffer, NbcA, NbcB, ClSize,
       }
-      cube_index += m_;
       private_index += tile_type::item_rows;
 
       slice_col += tile_type::wg_cols;
-      cube_index = slice_col * m_;
+      cube_index += cube_index_inc;
       private_index_offset += tile_type::item_rows;
     }
   }
 
   template <typename local_ptr_t>
   SYCL_BLAS_INLINE void extract_input_blocks(
-      index_t local_id, index_t tile_idx, local_ptr_t scratch_ptr,
-      local_ptr_t rhs_scratch_ptr, index_t global_m_offset,
-      index_t global_n_offset, index_t global_k_offset, bool check_m_limit,
-      bool check_n_limit) {
+      const index_t& lhs_row, const index_t& lhs_col, const index_t& rhs_row,
+      const index_t& rhs_col, const index_t& tile_idx, local_ptr_t scratch_ptr,
+      local_ptr_t rhs_scratch_ptr, const index_t& global_m_offset,
+      const index_t& global_n_offset, const index_t& global_k_offset,
+      bool check_m_limit, bool check_n_limit) {
     const bool check_k_limit =
-        global_k_offset + (tile_idx + 1) * tile_size_dim_k > k_;
+        IsFinal ? ((tile_idx + 1) * tile_size_dim_k > k_)
+                : (global_k_offset + (tile_idx + 1) * tile_size_dim_k > k_);
     const bool check_limits = check_m_limit || check_n_limit || check_k_limit;
     if (check_limits)
-      load_blocks<true, true, true>(local_id, tile_idx, scratch_ptr,
-                                    rhs_scratch_ptr, global_m_offset,
-                                    global_n_offset, global_k_offset);
+      load_blocks<true, true, true>(
+          lhs_row, lhs_col, rhs_row, rhs_col, tile_idx, scratch_ptr,
+          rhs_scratch_ptr, global_m_offset, global_n_offset, global_k_offset);
     else
-      load_blocks<false, false, false>(local_id, tile_idx, scratch_ptr,
-                                       rhs_scratch_ptr, global_m_offset,
-                                       global_n_offset, global_k_offset);
+      load_blocks<false, false, false>(
+          lhs_row, lhs_col, rhs_row, rhs_col, tile_idx, scratch_ptr,
+          rhs_scratch_ptr, global_m_offset, global_n_offset, global_k_offset);
   }
 
   template <bool check_m_limit, bool check_n_limit, bool check_k_limit,
             typename local_ptr_t>
-  SYCL_BLAS_INLINE void load_blocks(index_t local_id, index_t tile_idx,
-                                    local_ptr_t scratch_ptr,
-                                    local_ptr_t rhs_scratch_ptr,
-                                    index_t global_m_offset,
-                                    index_t global_n_offset,
-                                    index_t global_k_offset) {
+  SYCL_BLAS_INLINE void load_blocks(
+      const index_t& lhs_row, const index_t& lhs_col, const index_t& rhs_row,
+      const index_t& rhs_col, const index_t& tile_idx, local_ptr_t scratch_ptr,
+      local_ptr_t rhs_scratch_ptr, const index_t& global_m_offset,
+      const index_t& global_n_offset, const index_t& global_k_offset) {
     // LHS tile
     if (TransA) {
```

Comment:
if you move `if(TransA )` and `if(TransB)` to the load_block it reduces the size of code and make the code more readable. So TransA or TransB will be passed as a template parameter to loadblock and loadblock will check if(trans)

Reply:
Done.